sets/sets.go

```diff
@@ -0,0 +1,80 @@
+// Package sets is a minimal implementation of a generic set data structure.
+
+package sets
+
+import (
+	"fmt"
+	"sort"
+
+	"golang.org/x/exp/constraints"
+)
+
+// Set is a minimal set that takes only ordered types: any type that supports the
+// operators < <= >= >.
+type Set[T constraints.Ordered] struct {
+	items map[T]struct{}
+}
+
+// New returns an empty set with capacity size. The capacity will grow and shrink as a
+// stdlib map.
+func New[T constraints.Ordered](size int) *Set[T] {
+	return &Set[T]{items: make(map[T]struct{}, size)}
+}
+
+// From returns a set from elements.
+func From[T constraints.Ordered](elements ...T) *Set[T] {
+	s := New[T](len(elements))
+	for _, i := range elements {
+		s.items[i] = struct{}{}
+	}
+	return s
+}
+
+// String returns a string representation of s, ordered. This allows to simply pass a
+// sets.Set as parameter to a function that expects a fmt.Stringer interface and obtain
+// a comparable string.
+func (s *Set[T]) String() string {
+	return fmt.Sprint(s.OrderedList())
+}
+
+func (s *Set[T]) Size() int {
+	return len(s.items)
+}
+
+// OrderedList returns a slice of the elements of s, ordered.
+// TODO This can probably be replaced in Go 1.20 when a generics slice packages reaches
+// the stdlib.
+func (s *Set[T]) OrderedList() []T {
+	elements := make([]T, 0, len(s.items))
+	for e := range s.items {
+		elements = append(elements, e)
+	}
+	sort.Slice(elements, func(i, j int) bool {
+		return elements[i] < elements[j]
+	})
+	return elements
+}
+
+// Contains returns true if s contains item.
+func (s *Set[T]) Contains(item T) bool {
+	_, found := s.items[item]
+	return found
+}
+
+// Difference returns a set containing the elements of s that are not in x.
+func (s *Set[T]) Difference(x *Set[T]) *Set[T] {
+	result := New[T](max(0, s.Size()-x.Size()))
```

Comment:
Err... if `x` and `s` are disjoint the result will be equal to `s` so the actual max size it can reach is the size of `s` no matter what's the size of `x` is.

Reply:
Correct. We are allocating memory, and in this case we must allocate it before knowing how much we really need.

This errs on the side of allocating as little as possible, which is still better than not allocating anything. The map will then grow automatically with the inserts.

We could allocate the maximum size (`s.Size()`) and then release what is not used, but Go doesn't currently have a way to shrink memory allocated to a map (except playing GC tricks), see https://github.com/golang/go/issues/54766.

Last, we could just not preallocate anything. But it would have no positive impact compared to the current code, actually it would incur more allocations.

Or am I missing something?

Reply:
If x is larger than s, you're allocating 0.
If x is smaller than s, you're still likely to be under-allocating unless x is totally included in s which shouldn't be the general case.
So you're almost always under-allocating. Is it worth it? If you really want to avoid the growth, then pre-allocate s. For the current use case of having N input directories where N is small (<10), the waste is probably much better then the reallocation+move.

Reply:
> So you're almost always under-allocating. Is it worth it?

That is the point. It is always worth it, in the sense that it cannot do harm and sometimes it does good.
For sure, for the sizes we are talking about for cogito usage, as you point out, the difference is unmeasurable.
Still, there is no harm, so I just want to be sure I understand your point.

Reply:
I just don't think it's worth it.

I don't know how Go handles the growth but the usual way is to double the current allocation (when still under some limit) whenever you need more room and so you end up with over-allocation anyway in most cases.

Reply:
@odormond I understand your position. We will have to agree to disagree :-)